### ISSUE TITLE
Arrays of variable templates cannot deduce their lengths from pack expansion in the initializer

### ISSUE THREAD

**Issue by Mq-b** [llvm/llvm-project#79750]
```cpp
#include <iostream>
#include <type_traits>

template<std::size_t...values>
constexpr std::size_t array[]{ values... };

int main() {
    std::cout << std::is_same_v<decltype(::array<1, 2, 3, 4, 5>),  const std::size_t[]>;
}
```

[Output](https://godbolt.org/z/E8jTW1GsW):

```
1
```
Both gcc and clang have this result, but this is obviously not true, msvc outputs **0**.

gcc and clang consider `array<1, 2, 3, 4, 5>` to be the same type as `const std::size_t[]`.

---

msvc will only output 1 when faced with the following code.

```cpp
#include <iostream>
#include <type_traits>

template<std::size_t...values>
constexpr std::size_t array[]{ values... };

int main() {
    std::cout << std::is_same_v<decltype(::array<1, 2, 3, 4, 5>), const std::size_t[5]>;
}
```

msvc considers `array<1, 2, 3, 4, 5>` and `const std::size_t[5]` to be of the same type.

- ***I think msvc is right***.

**Comment by frederick-vs-ja** [comment]
Curiously, `sizeof(::array<1, 2, 3, 4, 5>)` correctly works for Clang (reports `40` for this case on an x64 platform).

**Comment by Mq-b** [comment]
> Curiously, `sizeof(::array<1, 2, 3, 4, 5>)` correctly works for Clang (reports `40` for this case on an x64 platform).

You're right, clang and gcc behave differently here, clang can `sizeof(array<1,2,3,4,5>)` and get 40.

If it is gcc, it is a compilation error.

```cpp
#include <iostream>
#include <type_traits>

template<std::size_t...values>
constexpr std::size_t array[]{ values... };

int main() {
    std::cout<<sizeof(array<1,2,3,4,5>);
}
```

https://godbolt.org/z/aMbGGzfGd

**Comment by shafik** [comment]
This does look like a bug but not 💯 sure

CC @erichkeane @Fznamznon

**Comment by beached** [comment]
Ran into this one today that seems related 

```cpp
enum class Values { A, B, C };

template<typename E, E...es>
struct test {
    static constexpr E values[] = {es...};
};

int main()
{
    using tst_t = test<Values, Values::A, Values::B, Values::C>;
    constexpr auto r = tst_t::values[0] == Values::A;
    return r;
}
```
https://compiler-explorer.com/z/r7sceTznc
```
<source>:11:20: error: constexpr variable 'r' must be initialized by a constant expression
   11 |     constexpr auto r = tst_t::values[0] == Values::A;
      |                    ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:11:24: note: read of element of array without known bound is not allowed in a constant expression
   11 |     constexpr auto r = tst_t::values[0] == Values::A;
      |                        ^
1 error generated.```

**Comment by beached** [comment]
I found this case, originally as a class static variable, but reduced to 

```cpp
enum class Values { A, B, C };


template<typename E>
inline constexpr Values values[] = {Values::A, Values::B, Values::C};


int main()
{
    constexpr auto r = values<Values>[0] == Values::A;
    return r;
}
```
Fails with ```
<source>:10:20: error: constexpr variable 'r' must be initialized by a constant expression
   10 |     constexpr auto r = values<Values>[0] == Values::A;
      |                    ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<source>:10:24: note: read of element of array without known bound is not allowed in a constant expression
   10 |     constexpr auto r = values<Values>[0] == Values::A;
      |                        ^
1 error generated.
Compiler returned: 1```

But  
```cpp
enum class Values { A, B, C };


inline constexpr Values values[] = {Values::A, Values::B, Values::C};


int main()
{
    constexpr auto r = values[0] == Values::A;
    return r;
}```

works.  Both work in msvc/gcc

**Comment by zyn0217** [comment]
GCC's behavior has been corrected since 13.3 https://godbolt.org/z/PeT687K9W

However, I have to admit my patch doesn't solve OP's case but only https://github.com/llvm/llvm-project/issues/79750#issuecomment-1965727668. I'll reopen the issue and look into it.

**Comment by zyn0217** [comment]
So this is interesting:

Clang relies on the variable being ODR used to instantiate its initializer, thereafter uses that definition to compute the complete type of the variable (i.e. `array[5]` instead of `array[]`). However since the use (`decltype(::array<1, 2, 3, 4, 5>)`) of variable in issue doesn't form an ODR-use, clang wouldn't instantiate the variable definition - hence it still considers it to be of an incomplete array type.

If we somehow make it use the variable, we'll be consistent with everyone else:

https://godbolt.org/z/4xv6frqnW

```cpp
#include <iostream>
#include <type_traits>

template<std::size_t...values>
constexpr std::size_t array[]{ values... };

int main() {
    auto x = array<1, 2, 3, 4, 5>;
    static_assert(std::is_same_v<decltype(::array<1, 2, 3, 4, 5>),  const std::size_t[5]>); // true!
}
```

I don't know if this should be a CWG issue @cor3ntin

**Comment by zygoloid** [comment]
Clang has a mechanism to handle this kind of issue -- look for calls to `RequireCompleteExprType` to see how this works. Presumably we should be calling that when computing the type for a `decltype` (or `__typeof`?) expression, but aren't.

**Comment by zyn0217** [comment]
> Clang has a mechanism to handle this kind of issue -- look for calls to `RequireCompleteExprType` to see how this works. Presumably we should be calling that when computing the type for a `decltype` (or `__typeof`?) expression, but aren't.


Thanks! I think I was just unsure if it's justified to instantiate initializers for such unevaluated operators.
